# **PRÀCTICA ÀLGEBRA**
Biel Fernandez Herencia

Izan Perez Sanchez

Edgar Saez Lopez

In [ ]:
import kagglehub

path = kagglehub.dataset_download("rajmehra03/movielens100k")


100%|██████████| 1.69M/1.69M [00:00<00:00, 49.3MB/s]

Extracting files...
Path to dataset files: /root/.cache/kagglehub/datasets/rajmehra03/movielens100k/versions/2


In [ ]:
import os
os.listdir(path)

['movies.csv', 'u.data', 'tags.csv', 'links.csv', 'ratings.csv']

In [ ]:
import pandas as pd

df = pd.read_csv(path+"/ratings.csv")
df.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [ ]:
df = df.iloc[12000:13000, :3]

df = pd.pivot_table(df, values="rating", index="userId",
                       columns="movieId")

df

movieId  1      2      6      7      10     16     20     21     25     32     \
userId                                                                          
75         NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
76         NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
77         4.0    NaN    3.5    NaN    3.0    NaN    NaN    NaN    NaN    4.0   
78         NaN    3.5    NaN    NaN    5.0    5.0    NaN    NaN    NaN    5.0   
79         2.0    2.5    NaN    0.5    NaN    NaN    NaN    NaN    NaN    NaN   
80         NaN    2.0    4.0    4.0    3.0    5.0    1.0    4.0    4.0    5.0   
81         NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
82         NaN    NaN    NaN    NaN    3.0    NaN    NaN    3.0    NaN    NaN   
83         NaN    NaN    5.0    NaN    NaN    NaN    NaN    1.0    NaN    NaN   

movieId  ...  79132  80463  81054  81591  81786  81834  82459  88125  88129  \
userId   ...                                                                  
75       ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
76       ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
77       ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
78       ...    4.5    NaN    NaN    NaN    NaN    4.5    NaN    4.5    4.0   
79       ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
80       ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
81       ...    3.0    3.0    4.0    2.5    4.0    NaN    3.5    NaN    NaN   
82       ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
83       ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

movieId  97673  
userId          
75         NaN  
76         NaN  
77         NaN  
78         4.5  
79         NaN  
80         NaN  
81         NaN  
82         NaN  
83         NaN  

[9 rows x 729 columns]

Les dimensions de la matriu son 9 x 729 ja que hi ha només 9 usuaris diferents en el rang de 12000 a 12999. Cada fila representa un usuari, i les columnes representen les diferents películes que han rebut valoracions.

In [ ]:
df = df.fillna(0)
df

movieId  1      2      6      7      10     16     20     21     25     32     \
userId                                                                          
75         0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
76         0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
77         4.0    0.0    3.5    0.0    3.0    0.0    0.0    0.0    0.0    4.0   
78         0.0    3.5    0.0    0.0    5.0    5.0    0.0    0.0    0.0    5.0   
79         2.0    2.5    0.0    0.5    0.0    0.0    0.0    0.0    0.0    0.0   
80         0.0    2.0    4.0    4.0    3.0    5.0    1.0    4.0    4.0    5.0   
81         0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
82         0.0    0.0    0.0    0.0    3.0    0.0    0.0    3.0    0.0    0.0   
83         0.0    0.0    5.0    0.0    0.0    0.0    0.0    1.0    0.0    0.0   

movieId  ...  79132  80463  81054  81591  81786  81834  82459  88125  88129  \
userId   ...                                                                  
75       ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
76       ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
77       ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
78       ...    4.5    0.0    0.0    0.0    0.0    4.5    0.0    4.5    4.0   
79       ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
80       ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
81       ...    3.0    3.0    4.0    2.5    4.0    0.0    3.5    0.0    0.0   
82       ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
83       ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

movieId  97673  
userId          
75         0.0  
76         0.0  
77         0.0  
78         4.5  
79         0.0  
80         0.0  
81         0.0  
82         0.0  
83         0.0  

[9 rows x 729 columns]

In [ ]:
import numpy as np
from numpy.linalg import svd

U, sigma, Vt = svd(df, full_matrices=False)

# construim la matriu diagonal perque la funcion svd el valor sigma es un array amb els valors singulars
Sigma = np.diag(sigma)

# Reconstruim la matriu pero substituint els valors 0 per les prediccions
pred = np.dot(np.dot(U, Sigma), Vt)

pred

np.float64(4.566557985116957e-16)

In [22]:
sigma

array([78.28443425, 53.36692855, 50.3734961 , 39.47853474, 23.05393457,
       22.21283523, 18.48505578, 17.14184453, 16.21081143])